File: python/searchable-analysis/src/searchable_analysis/stemmer_scandinavian.py
def mark_scandinavian_region(word: str, vowels: frozenset[str]) -> int:
    index = 0
    while index < len(word) and word[index] not in vowels:
        index += 1
    while index < len(word) and word[index] in vowels:
        index += 1
    if index < len(word):
        index += 1
    return max(3, index)


def longest_suffix_in_region(word: str, region_start: int, suffixes: tuple[str, ...]) -> str | None:
    matches = [suffix for suffix in suffixes if len(word) - len(suffix) >= region_start and word.endswith(suffix)]
    return max(matches, key=len, default=None)

File: packages/searchable/src/analysis/stemmer-scandinavian.ts
export function markScandinavianRegion(
  word: string,
  vowels: ReadonlySet<string>,
): number {
  let index = 0;
  while (index < word.length && !vowels.has(word[index] as string)) index++;
  while (index < word.length && vowels.has(word[index] as string)) index++;
  if (index < word.length) index++;
  return Math.max(3, index);
}

export function longestSuffixInRegion(
  word: string,
  regionStart: number,
  suffixes: readonly string[],
): string | undefined {
  let best: string | undefined;
  for (const suffix of suffixes) {
    const start = word.length - suffix.length;
    if (
      start >= regionStart &&
      word.endsWith(suffix) &&
      (best === undefined || suffix.length > best.length)
    ) {
      best = suffix;
    }
  }
  return best;
}
